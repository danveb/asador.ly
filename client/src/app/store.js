import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../redux/auth/authSlice";
import pinReducer from "../redux/pins/pinSlice";

export const store = configureStore({
  reducer: {
    auth: authReducer,
    pins: pinReducer,
  },
}); 