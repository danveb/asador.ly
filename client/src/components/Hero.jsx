import { Link } from "react-router-dom"; 
import { Footer } from "./index"; 
import brasa from "../assets/about/brasa.png"; 
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import "../styles/Hero.scss"; 

const Hero = () => {
    return (
        <div>
            <div className="hero">
                <div className="hero-top">
                    <div className="hero-title">
                        <h3>Explore...</h3>
                        <h1>Buenos Aires</h1>
                        <Link to="/map">Check Our Map</Link>
                    </div>
                    <div className="hero-img">
                        <img src={brasa} alt="charcoal" />
                    </div>
                    <div className="arrow">
                        <KeyboardArrowDownIcon />
                    </div>
                </div>
                <div className="hero-bottom">
                    <div className="hero-subtitle">
                        <h2>Asador</h2>
                        <p>It's all about sharing the joy of <span>asado</span> (barbeque) that  makes Argentina a prime destination for dining at <span>parrillas</span> (steakhouses). From traditional eateries to fine-dining establishments, there's always a <span>parrilla</span> to choose from for a fix of authentic Argentine <span>asado</span>. </p>
                        <p>Here at <span>Asador</span> we have curated a list we recommend of the best steakhouses in Buenos Aires (we may have missed a few, on purpose, just so you can share your experience with us too) for every budget. Carnivores rejoice!</p> 
                    </div>
                </div>
            </div>
            <Footer />
        </div>
    )
}

export default Hero