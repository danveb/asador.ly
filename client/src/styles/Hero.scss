.hero {
    width: 100%; 
    display: flex; 
    flex-direction: column; 
    font-family: "Prata", sans-serif; 
    color: #fff; 

    .hero-top {
        height: calc(100vh - 55px); 

        .hero-title {
            position: absolute; 
            left: 3%; 
            top: 35%; 

            h3 {
                font-size: 30px; 
                color: #a67d4a; 

                @media screen and (max-width: 480px) {
                    font-size: 22px; 
                }
            }

            h1 {
                font-size: 100px; 
                color: inherit;
                margin-bottom: 20px; 
                font-weight: 500; 

                @media screen and (max-width: 768px) {
                    font-size: 70px; 
                }

                @media screen and (max-width: 480px) {
                    font-size: 50px; 
                }
            }

            a {
                padding: 15px; 
                color: #fff; 
                text-decoration: none; 
                cursor: pointer; 
                border: 2px solid rgb(0, 173, 230); 
                border-image: linear-gradient(to right, rgb(0, 173, 230) 0%, rgb(210, 224, 252) 100%) 2 / 1 / 0 stretch;
                font-family: "Poppins", sans-serif; 
            }
        }

        .hero-img {
            height: 100%; 

            img {
                width: 100%; 
                height: 100%; 
                object-fit: cover; 
            }
        }

        .arrow {
            width: 100%; 
            position: absolute; 
            bottom: 25px; 
            text-align: center; 

            svg {
                width: 30px; 
                height: 30px; 
            }
        }
    }

    .hero-bottom {
        width: 100%; 
        padding: 40px 24px; 
        background-color: #fff; 
        color: #000; 
        font-weight: 100; 

        .hero-subtitle {
            max-width: 1600px; 
            width: 50%; 
            margin: 0 auto; 
            display: flex; 
            justify-content: center; 
            flex-direction: column; 

            @media screen and (max-width: 768px) {
                width: 75%; 
            }

            @media screen and (max-width: 480px) {
                width: 100%; 
            }

            h2 {
                font-family: "Prata", sans-serif; 
                font-size: 42px; 
                margin-bottom: 25px; 

                @media screen and (max-width: 768px) {
                    font-size: 30px; 
                }
            }

            p {
                font-family: "Yrsa", sans-serif; 
                font-size: 18px; 
                line-height: 32px; 
                margin-bottom: 25px; 

                span {
                    font-style: italic; 
                }
            }
        }
    }
}