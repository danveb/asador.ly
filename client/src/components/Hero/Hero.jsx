import { useEffect, useState } from "react"; 
import Spinner from "../Spinner/Spinner";
import Map, { Marker, Popup } from "react-map-gl";
import RoomIcon from '@mui/icons-material/Room';
import StarRateIcon from '@mui/icons-material/StarRate';
import AttachMoneyIcon from '@mui/icons-material/AttachMoney';
import { useSelector, useDispatch } from "react-redux"; 
import { createPin, getPins, reset } from "../../redux/pins/pinSlice"; 
import "./Hero.scss"; 
import { toast } from "react-toastify";

const Hero = () => {
    const { user } = useSelector((state) => state.auth); 
    const { pins, isLoading, isError, message } = useSelector((state) => state.pins); 

    // useDispatch
    const dispatch = useDispatch(); 

    // useState
    const [viewState, setViewState] = useState({
        // main coordinates for center view
        longitude: -58.403578,
        latitude: -34.608056,
        zoom: 11.5
    });
    const [showPopup] = useState(true); 
    const [currentPlaceId, setCurrentPlaceId] = useState(null); 
    const [newPlace, setNewPlace] = useState(null); 
    const [formData, setFormData] = useState({
        title: "", 
        description: "", 
        rating: "", 
        cost: "", 
    });

    const { title, description, rating, cost } = formData; 

    const handleChange = (e) => {
        setFormData((prevState) => ({
            ...prevState, 
            [e.target.name]: e.target.value
        })); 
    }; 

    // useEffect
    useEffect(() => {
        if(isError) {
            console.log(message); 
        }
        dispatch(getPins()); 
    }, [dispatch, isError, message]); 

    // handleMarkerClick
    const handleMarkerClick = (id, long, lat) => {
        setCurrentPlaceId(id); 
        setViewState({
            ...viewState, 
            longitude: long, 
            latitude: lat, 
        }); 
    }; 

    // handleAddClick 
    const handleAddClick = (e) => {
        if(user) {
            const latitude = e.lngLat.lat; 
            const longitude = e.lngLat.lng; 
            setNewPlace({
                lat: latitude,
                lng: longitude, 
            }); 
        } else {
            toast("Please consider creating an account to add a new place."); 
        }; 
    }; 

    // handleSubmit (REDUX) 
    const handleSubmit = (e) => {
        e.preventDefault(); 
        const pinData = {
            username: user.username, 
            title,
            description, 
            rating, 
            cost, 
            location: {
                coordinates: [newPlace.lng, newPlace.lat] 
            }, 
        }; 
        console.log(pinData); 
        dispatch(createPin(pinData)); 
        dispatch(reset()); 
    }; 

    if(isLoading) {
        return <Spinner />
    }; 
    
    return (
        <div className="hero">
            <Map
                // controlled map (state)
                {...viewState} 
                onMove={evt => setViewState(evt.viewState)}
                style={{width: "100%", height: "100%"}}
                // mapStyle="mapbox://styles/mapbox/streets-v9"
                // custom map style
                mapStyle="mapbox://styles/danvebb/cl6frln64000414mrhya0xc3k"
                // mapbox token
                mapboxAccessToken={process.env.REACT_APP_MAPBOX}
                // double click to create a pin 
                onDblClick={handleAddClick}
                transition="300"
            >
                {pins.map((pin, idx) => (
                    <div key={idx}>
                        <Marker 
                            longitude={pin.location.coordinates[0]} 
                            latitude={pin.location.coordinates[1]} 
                            anchor="bottom" 
                            offsetLeft={-viewState.zoom * 3.5}
                            offsetTop={-viewState.zoom * 7}
                        >
                            <RoomIcon 
                                style={{ color: "crimson", fontSize: viewState.zoom * 4, cursor: "pointer" }}
                                onClick={() => handleMarkerClick(pin._id, pin.location.coordinates[0], pin.location.coordinates[1])}
                            />
                        </Marker>
                        {showPopup && pin._id === currentPlaceId && (
                            <Popup 
                                longitude={pin.location.coordinates[0]} 
                                latitude={pin.location.coordinates[1]}
                                anchor="top"
                                closeOnClick={false}
                                onClose={() => setCurrentPlaceId(null)}
                            >
                                <div className="card">
                                    <label>Place</label>
                                    <h4 className="place">{pin.title}</h4>
                                    <label>Description</label>
                                    <p className="description">{pin.description}</p>
                                    <label>Rating</label>
                                    <div className="icons">
                                        {Array(pin.rating).fill(<StarRateIcon className="star" />)}
                                    </div>
                                    <label>Cost</label>
                                    <div className="icons">
                                        {Array(pin.cost).fill(<AttachMoneyIcon className="dollar" />)}
                                    </div>
                                    <label>Info</label>
                                    <span className="username">Created by <b>{pin.username}</b></span>
                                </div>
                            </Popup> 
                        )}
                    </div>
                ))}
                {showPopup && newPlace && (
                    <Popup 
                        longitude={newPlace.lng} 
                        latitude={newPlace.lat}
                        anchor="top"
                        closeOnClick={false}
                        onClose={() => setNewPlace(null)}
                    >
                        <div>
                            <form onSubmit={handleSubmit}>
                                <label 
                                    htmlFor="title"
                                >Title</label>
                                <input 
                                    id="title"
                                    name="title"
                                    value={title}
                                    placeholder="Enter Title" 
                                    onChange={handleChange}
                                    required
                                />
                                <label 
                                    htmlFor="description"
                                >Description</label>
                                <textarea 
                                    id="description"
                                    name="description"
                                    value={description}
                                    placeholder="Want To Share About This Steakhouse?"
                                    onChange={handleChange}
                                />
                                <label 
                                    htmlFor="rating"
                                >Rating</label>
                                <select 
                                    id="rating"
                                    name="rating"
                                    value={rating}
                                    onChange={handleChange}
                                >
                                    <option defaultValue={"default"}>Choose an option</option>
                                    <option value={"1"}>1</option>
                                    <option value={"2"}>2</option>
                                    <option value={"3"}>3</option>
                                    <option value={"4"}>4</option>
                                    <option value={"5"}>5</option>
                                </select>
                                <label 
                                    htmlFor="cost"
                                >Cost</label>
                                <select 
                                    id="cost"
                                    name="cost"
                                    value={cost}
                                    onChange={handleChange}
                                >
                                    <option defaultValue={"default"}>Choose an option</option>
                                    <option value={"1"}>1</option>
                                    <option value={"2"}>2</option>
                                    <option value={"3"}>3</option>
                                    <option value={"4"}>4</option>
                                    <option value={"5"}>5</option>
                                </select>
                                <button className="submit-btn" type="submit">Add Pin</button>
                            </form>
                        </div>
                    </Popup>
                )}
            </Map>
        </div>
    )
}

export default Hero