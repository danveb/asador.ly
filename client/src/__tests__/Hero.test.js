import { render } from "@testing-library/react"; 
import { store } from "../app/store"; 
import { Provider } from "react-redux";
import Hero from "../components/Hero/Hero"; 

describe("Hero component", () => {
    // smoke test
    test("renders Hero component", () => {
        render(
            <Provider store={store}>
                <Hero />
            </Provider>
        );
    });

    // snapshot test
    test("matches snapshot", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Hero />
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot(); 
    });
});