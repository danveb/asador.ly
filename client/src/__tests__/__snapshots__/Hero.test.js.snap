// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner-container"
  >
    <div
      class="spinner-loading"
    />
  </div>
</DocumentFragment>
`;
